specs/agents/tracing-instrumentation-aws-lambda.md

```diff
@@ -0,0 +1,143 @@
+`service.framework.name` | `AWS Lambda` | Constant value for the framework name. | -
+`service.runtime.name`| e.g. `AWS_Lambda_java8` | The lambda runtime. | `AWS_EXECUTION_ENV`
+`service.id` | e.g. `arn:aws:lambda:us-west-2:123456789012:function:my-function` | The ARN of the function **without alias suffix**. | `context.invokedFunctionArn`, remove the 8th ARN segment if the ARN contains an alias suffix. `arn:aws:lambda:us-west-2:123456789012:function:my-function:someAlias` will become `arn:aws:lambda:us-west-2:123456789012:function:my-function`.
+`service.version` | e.g. `${LATEST}` | The lambda function version | `AWS_LAMBDA_FUNCTION_VERSION` or `context.functionVersion`
+`service.node.name` | e.g. `2019/06/07/[$LATEST]e6f...` | The log stream name uniquely identifying a function instance. | `AWS_LAMBDA_LOG_STREAM_NAME` or `context.logStreamName`
+`cloud.provider` | `aws` | Constant value for the cloud provider. | 
+`cloud.origin.provider` | `aws` | Constant value for the origin cloud provider. | 
+`cloud.region` | e.g. `us-east-1` | The cloud region. | `AWS_REGION`
+`cloud.service.name` | `lambda` |  Constant value for the AWS service.
+`transaction.type` | `request` | - | -
+`faas.trigger.type` | `other` | The trigger type. Use `other` if trigger type is unknown / cannot be specified. | More concrete triggers are `http`, `pubsub`, `datasource`, `timer` (see specific triggers below). 
+`faas.execution` | `af9aa4-a6...` | The AWS request ID of the function invocation | `context.awsRequestId`
+`faas.coldstart` | `true` / `false` | Boolean value indicating whether a Lambda function invocation was a cold start or not. | [see section below](deriving-cold-starts)
+`transaction.name` | e.g. `MyFunctionName` | Use function name if trigger type is `other`. | `context.functionName` 
+`faas.trigger.reuqest_id` | - | Do not set this field if trigger type is `other`.  | Trigger specific.
+`service.origin.*` | - | Do not set these fields if trigger type is `other`. | Trigger specific. 
+`cloud.origin.*` | - | Do not set these fields if trigger type is `other`.  | Trigger specific.  
+
+### Overwriting Metadata
+Automatically capturing cloud metadata doesn't work reliably from a Lambda environment. Moreover, retrieving cloud metadata through an additional HTTP request may slowdown the lambda function / increase cold start behaviour. Therefore, the generic cloud metadata fetching should be disabled when the agent is running in a lambda context (for instance through checking for the existance of the `AWS_LAMBDA_FUNCTION_NAME` environment variable). 
+Where possible, metadata should be overwritten at Lambda runtime startup corresponding to the field specifications in this spec.
+Some metadata fields are not available at startup (e.g. `invokedFunctionArn`). In this case corresponding fields need to be set either as metadata with the first invocation of the lambda function or set for every lambda invocation as transaction fields.
```

Comment:
I'm a little unclear on whether some of these fields need to be set on the metadata only, the transaction only, if the intention is that all these fields will eventually be added to _both_ the metadata and transaction objects, or if we're meant to pick where a field goes from context that up from context.  

Using the current [APM Server schemas](https://github.com/elastic/apm-server/tree/master/docs/spec/v2) as a guide, we have

Fields that don't exist on either the metadata or the transaction

|Prop |Current Metadata |Current Transaction|
|-----|-----|-----|
|cloud.origin.account.id | NO | NO|
|cloud.origin.provider | NO | NO|
|cloud.origin.region | NO | NO|
|cloud.origin.service.name | NO | NO|
|faas.coldstart | NO | NO|
|faas.execution | NO | NO|
|faas.trigger.reuqest_id | NO | NO|
|faas.trigger.type | NO | NO|
|service.node.name | NO | NO|
|service.id | NO | NO|
|service.origin | NO | NO|
|service.origin.id | NO | NO|
|service.origin.name | NO | NO|
|service.origin.version | NO | NO|

Fields that are metadata only

|Prop |Current Metadata |Current Transaction|
|-----|-----|-----|
|cloud.provider | YES | NO|
|cloud.region | YES | NO|
|cloud.service.name | YES | NO|


Fields that are transaction only

|Prop |Current Metadata |Current Transaction|
|-----|-----|-----|
|transaction.name | NO | YES|
|transaction.type | NO | YES|
|message.age | NO | YES|
|message.body | NO | YES|
|message.headers | NO | YES|
|message.queue | NO | YES|

and fields that exist on both the metadata and the transaction

|Prop |Current Metadata |Current Transaction|
|-----|-----|-----|
|service.framework.name | YES | YES|
|service.name | YES | YES|
|service.runtime.name | YES | YES|
|service.version | YES | YES|

Reply:
> Fields that don't exist on either the metadata or the transaction

Fields that are not operation-specific belong in metadata.

> service.node.name

This is called `service.node.configured_name` in the intake API; it is recorded as `service.node.name` in Elasticsearch docs.

> service.id

This should be added to metadata.

I think everything else you listed should be transaction-only.